### PR TITLE
Stream

### PR DESCRIPTION
**Pull request by ZhiHanZ**
Makefile and regex matching

### CODE REVIEW

File: utils/regex.cc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  std::sregex_iterator next(subject.begin(), subject.end(), re);
+  std::sregex_iterator end;
+  vector<string> tags;
+  while (next != end) {
+    std::smatch match = *next;
+    tags.push_back(match.str());
+    next++;
+  }
+  return tags;
+}  // namespace utils
```
Comment (barath) [inline utils/regex.cc]: This isn't the end of the namespace.

Comment (ZhiHanZ) [reply utils/regex.cc]: fixed comment error

File: utils/regex.cc [hunk 2]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (barath) [inline utils/regex.cc]: Remove blank line.

File: utils/regex.cc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#include "regex.h"
+#include <iostream>
+#include <regex>
+#include <string>
+#include <vector>
```
Comment (barath) [inline utils/regex.cc:5]: Add blank line between includes and using statements.

File: utils/regex.cc [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using std::regex;
+using std::smatch;
+using std::sregex_iterator;
+using std::vector;
```
Comment (barath) [inline utils/regex.cc:10]: Add blank line.

File: utils/regex.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#ifndef UTILS_REGEX_H_
+#define UTILS_REGEX_H_
+
+#include <string>
+#include <vector>
+using std::string;
```
Comment (barath) [inline utils/regex.h:6]: Add blank lines where needed.

File: utils/regex.cc [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace utils {
+// find all hashtags in string subject and put them into a vector
```
Comment (barath) [inline utils/regex.cc]: Remove comment from cc file.

File: utils/regex.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+using std::vector;
+namespace utils {
+// extract all hashtags from subject and put them into vector
```
Comment (barath) [inline utils/regex.h]: Rephrase comment to describe the function in terms of its argument and return value more explicitly.

File: utils/regex_test.cc [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#include "regex.h"
+#include <gtest/gtest.h>
+#include <string>
+#include <vector>
+using std::string;
+using std::vector;
+using utils::HashParser;
+// test basic cases in HashParser
```
Comment (barath) [inline utils/regex_test.cc]: Make test comments more descriptive.

File: utils/regex.cc (hunk 2 above)
Comment (ZhiHanZ) [reply utils/regex.cc]: fixed

File: utils/regex.cc (hunk 3 above)
Comment (ZhiHanZ) [reply utils/regex.cc:5]: fixed

File: utils/regex.cc (hunk 4 above)
Comment (ZhiHanZ) [reply utils/regex.cc:10]: fixed

File: utils/regex.h (hunk 5 above)
Comment (ZhiHanZ) [reply utils/regex.h:6]: fixed

File: utils/regex.cc (hunk 6 above)
Comment (ZhiHanZ) [reply utils/regex.cc]: fixed

File: utils/regex.h (hunk 7 above)
Comment (ZhiHanZ) [reply utils/regex.h]: fixed

File: utils/regex_test.cc (hunk 8 above)
Comment (ZhiHanZ) [reply utils/regex_test.cc]: add more precise comments

**Review by barath — APPROVED** [review]

**Review by Stephanielpt — APPROVED** [review]

**Pull request merged**